gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/fusion/Fusion.java

```diff
@@ -0,0 +1,187 @@
+      //final String label = FusionUtils.getGraphLabel(patternGraph);
+
+      // Then I create the graph that substitute the vertices within toBeReplaced
+      DataSet<Vertex> toBeAdded;
+
+      toBeAdded = searchGraph.getGraphHead().first(1).join(patternGraph.getGraphHead().first(1))
+        .where((GraphHead g) -> 0).equalTo((GraphHead g) -> 0)
+        .with(new FlatJoinFunction<GraphHead, GraphHead, Vertex>() {
+          @Override
+          public void join(GraphHead searchGraphHead, GraphHead patternGraphSeachHead,
+            Collector<Vertex> out) throws Exception {
+            Vertex v = new Vertex();
+            v.setLabel(patternGraphSeachHead.getLabel());
+            v.setProperties(patternGraphSeachHead.getProperties());
+            v.setId(vId);
+            v.addGraphId(searchGraphHead.getId());
+            out.collect(v);
+          }
+        });
+
+      /*
+       * The newly created vertex v has to be created iff. we have some actual vertices to be
+       * replaced, and then if toBeReplaced contains at least one element
+       */
+      DataSet<Vertex> addOnlyIfNecessary =
+        toBeReplaced.first(1).join(toBeAdded).where((Vertex x) -> 0).equalTo((Vertex x) -> 0)
+          .with(new JoinFunction<Vertex, Vertex, Vertex>() {
+            @Override
+            public Vertex join(Vertex first, Vertex second) throws Exception {
+              // If there is at least one element to be replaced, then return the vertex that has
+              // to be added. Otherwise, this function won't ever be called
+              return second;
+            }
+          });
+
+      DataSet<Vertex> toBeReturned = finalVertices.union(addOnlyIfNecessary);
+      //////////////////////////////////////////
+
+      //In the final graph, all the edges appearing only in the search graph should appear
+      DataSet<Edge> leftEdges = searchGraph.getEdges();
+      leftEdges = FusionUtils.areElementsInGraph(leftEdges, patternGraph, false);
+
+            /*
+             * Concerning the other edges, we have to eventually update them and to be linked
+             * with the new vertex
+             * The following expression could be formalized as follows:
+             *
+             * updatedEdges = map(E, x ->
+             *      e' <- onNextUpdateof(x).newfrom(x);
+             *      if (e'.src \in toBeReplaced) e'.src = vId
+             *      end if
+             *      if (e'.dst \in toBeReplaced) e'.dst = vId
+             *      end if
+             *      return e'
+             * )
+             *
+             */
+      DataSet<Edge> updatedEdges =
+        leftEdges.fullOuterJoin(toBeReplaced).where((Edge x) -> x.getSourceId())
+          .equalTo((Vertex y) -> y.getId())
+          .with((FlatJoinFunction<Edge, Vertex, Edge>) (edge, vertex, collector) -> {
+              if (vertex == null) {
+                collector.collect(edge);
+              } else if (edge != null) {
+                Edge e = new Edge();
```

Comment:
Just one thing. The logic is dual, so I'm using a boolean variable to distinguish the two cases.